docs/CODING_CONVENTIONS.md

```diff
@@ -0,0 +1,246 @@
+
+This prevents consumers of your API from relying on pattern-matching which is not future proof against added error
+variants.
+
+###### Example
+
+Implementing a non-exhaustive variant is done like this:
+
+```rust
+pub enum Error {
+    PermissionDenied,
+    NotConnected,
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+```
+
+It informs users of your API that they should include a catch-all match when matching against the error like this:
+
+```rust
+match e {
+    PermissionDenied => { /*  */ },
+    NotConnected => { /*  */ },
+    _ => { /*  */ },
+}
+```
+
+This in turn guards any matches from being non-exhaustive in case new error variants are added in the future.
+
+#### Use `amethyst_error::Error` for compositional APIs
+
+APIs which composes results from multiple crates _should_ use `amethyst_error::Error`.
+This is a generic error type which is capable of boxing any error and annotate it with debugging information.
+
+This must be used when defining APIs which composes errors generically, like with traits.
+This is also required when composing errors from other crates, since It is currently the only mechanism available to
+communicate backtraces.
+
+###### Do
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    use amethyst_error::Error;
+
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, Error> {
+            Err(Error::from(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// no dependencies
+mod c {
+    use amethyst_error::Error;
+
+    pub trait Example {
+        fn foo() -> Result<u32, Error>;
+    }
+}
+```
+
+###### Don't
+
+```rust
+// crate `a`
+// depends on `c`
+mod a {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::A(error::Error::Problem))
+        }
+    }
+}
+
+// crate `b`
+// depends on `c`
+mod b {
+    pub mod error {
+        pub enum Error {
+            Problem,
+        }
+    }
+
+    struct Example;
+
+    impl c::Example for Example {
+        fn foo() -> Result<u32, c::Error> {
+            Err(c::Error::B(error::Error::Problem))
+        }
+    }
+}
+
+// crate `c`
+// depends on `a` and `b`.
+mod c {
+    use crate::{a, b};
+
+    pub enum Error {
+        A(a::error::Error),
+        B(b::error::Error),
+    }
+
+    pub trait Example {
+        fn foo() -> Result<u32, Error>;
+    }
+}
+```
+
+#### Do not overload the default `Result`
```

Comment:
Very good point, I have not thought about this and I will have to refactor some of my crates 👍

Reply:
me too :smile: